File: src/hash/ThreeSum.java
package hash;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int target = -nums[i];
            int l = i + 1, r = nums.length - 1;
            while (r > l) {
                if (nums[l] + nums[r] + nums[i] > 0) {
                    r--;
                } else if (nums[l] + nums[r] + nums[i] < 0) {
                    l++;
                } else {
                    List<Integer> list = new ArrayList<>();
                    list.add(nums[i]);
                    list.add(nums[l]);
                    list.add(nums[r]);
                    ans.add(list);
                    while (l < r && nums[l] == nums[l + 1]) l++;
                    while (l < r && nums[r] == nums[r - 1]) r--;

                    r--;
                    l++;
                }

            }

        }
        return ans;

    }

}

File: LeetcodeKt/src/twopoints/ThreeSum.kt
package twopoints

/*15. 三数之和
 */

class ThreeSum {
    fun threeSum(nums: IntArray): List<List<Int>> {
        val ans = mutableListOf<List<Int>>()
        val n = nums.size
        nums.sort()
        for (first in nums.indices) {
            if (first > 0 && nums[first] == nums[first - 1]) {
                continue
            }
            var third = n - 1
            val target = -nums[first]
            for (second in (first + 1) until n) {
                if (second > first + 1 && nums[second] == nums[second - 1]) {
                    continue
                }

                while (second < third && nums[second] + nums[third] > target) {
                    third--
                }

                if (second == third) {
                    break
                }
                if (nums[second] + nums[third] == target) {
                    val list = mutableListOf<Int>()
                    list.add(nums[first])
                    list.add(nums[second])
                    list.add(nums[third])
                    ans.add(list)
                }
            }
        }
        return ans
    }
}